File: easy/1275. Find Winner on a Tic Tac Toe Game.java
class Solution {
    public String tictactoe(int[][] moves) {
        /*
        1. 행, 열, 대각선 중 3개가 동일한 문자이면 승리
           - 가능한 행의 갯수 = 3개
           - 가능한 열의 갯수 = 3개
           - 가능한 대각선의 갯수 = 2개
           => 한 행 또는 한 열 또는 한 대각선을 기준으로 3칸을 갖고 있으면 됨
        2. 승부가 나지 않았는데, 사각형이 꽉찼으면 draw
        3. 승부가 나지 않았고, 사각형이 비어있으면 pending
        */
        
        //각 플레이어의 각 행, 열, 대각선에 위치하는 해당 플레이어의 문자열 갯수를 저장
        int[] A = new int[8], B = new int[8];
        
        for(int i=0;i<moves.length; i++){
            int y = moves[i][0], x = moves[i][1];
            int[] player = i%2==0 ? A : B;
            player[y]++;
            player[x+3]++;
            if(y==x) player[6]++;
            if(y==2-x) player[7]++;
        }
        
        for(int i=0; i<8;i++){
            if(A[i]==3) return "A";
            else if (B[i]==3) return "B";
        }
        
        if(moves.length == 9) return "Draw";
        else return "Pending";
        
    }
    
    //시간 복잡도 : O(n+8) -> O(n)
    //공간 복잡도 : O(16) -> O(1)
    
}

File: easy/1275. Find Winner on a Tic Tac Toe Game.py
#TC(mn) : m is the len(moves), n : length of board -> '3' in this question
#SC(n^2) 

class Solution:
    def tictactoe(self, moves: List[List[int]]) -> str:
        n = 3
        board = [[0 for i in range(n)] for j in range(n)]

        for x in range(len(moves)) :
            #check who's win everytime when the new input is came.
            row = moves[x][0]
            col = moves[x][1]
            if x%2 == 0 :
                board[row][col] = 'X'
                if self.didWin(board, row, col, 'X', n) : return 'A'
            else :
                board[row][col] = 'O'
                if self.didWin(board, row, col, 'O', n) : return 'B'
            
        if len(moves) == n*n :
            return "Draw"
        else : 
            return "Pending"
            
    def didWin(self, board, row, col, player, n) :
        didPlayerwin = True
        for x in range(n) :
            if board[row][x] != player :
                didPlayerwin = False
        
        if didPlayerwin : return True
        
        didPlayerwin = True
        
        for x in range(n) :
            if board[x][col] != player :
                didPlayerwin = False
        
        if didPlayerwin : return True
        
        if row == col : 
            didPlayerwin = True
            for x in range(n) :
                if board[x][x] != player :
                    didPlayerwin = False
            if didPlayerwin : return True
        
        if col == n-row-1 :
            didPlayerwin = True
            for x in range(n) :
                if board[x][n-x-1] != player :
                    didPlayerwin = False
            if didPlayerwin : return True
        
        return False
